service/header/store.go

```diff
@@ -0,0 +1,302 @@
+}
+
+func NewStore(ds datastore.Batching) (Store, error) {
+	return newStore(ds)
+}
+
+func NewStoreWithHead(ds datastore.Batching, head *ExtendedHeader) (Store, error) {
+	store, err := newStore(ds)
+	if err != nil {
+		return nil, err
+	}
+
+	err = store.put(head)
+	if err != nil {
+		return nil, err
+	}
+
+	return store, store.newHead(head.Hash())
+}
+
+func newStore(ds datastore.Batching) (*store, error) {
+	ds = namespace.Wrap(ds, storePrefix)
+	cache, err := lru.NewARC(DefaultStoreCache)
+	if err != nil {
+		return nil, err
+	}
+
+	index, err := newHeightIndexer(ds)
+	if err != nil {
+		return nil, err
+	}
+
+	return &store{
+		ds:    ds,
+		cache: cache,
+		index: index,
+	}, nil
+}
+
+func (s *store) Start(context.Context) error {
+	err := s.loadHead()
+	if err == datastore.ErrNotFound {
+		return fmt.Errorf("header/store: no chain head")
+	}
+
+	return err
+}
+
+func (s *store) Head(ctx context.Context) (*ExtendedHeader, error) {
+	s.headLk.RLock()
+	head := s.head
+	s.headLk.RUnlock()
+	return s.Get(ctx, head)
+}
+
+func (s *store) Get(_ context.Context, hash bytes.HexBytes) (*ExtendedHeader, error) {
+	if v, ok := s.cache.Get(hash); ok {
+		return v.(*ExtendedHeader), nil
+	}
+
+	b, err := s.ds.Get(datastore.NewKey(hash.String()))
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return nil, ErrNotFound
+		}
+
+		return nil, err
+	}
+
+	return UnmarshalExtendedHeader(b)
+}
+
+func (s *store) GetByHeight(ctx context.Context, height int64) (*ExtendedHeader, error) {
+	hash, err := s.index.HashByHeight(height)
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return nil, ErrNotFound
+		}
+
+		return nil, err
+	}
+
+	return s.Get(ctx, hash)
+}
+
+func (s *store) GetRangeByHeight(ctx context.Context, from, to int64) ([]*ExtendedHeader, error) {
+	h, err := s.GetByHeight(ctx, to)
+	if err != nil {
+		return nil, err
+	}
+
+	ln := to - from
+	headers := make([]*ExtendedHeader, ln)
+	for i := ln - 1; i > 0; i-- {
+		headers[i] = h
+		h, err = s.Get(ctx, h.LastHeader())
+		if err != nil {
+			return nil, err
+		}
+	}
+	headers[0] = h
+
+	return headers, nil
+}
+
+func (s *store) Has(_ context.Context, hash bytes.HexBytes) (bool, error) {
+	if ok := s.cache.Contains(hash); ok {
+		return ok, nil
+	}
+
+	return s.ds.Has(datastore.NewKey(hash.String()))
+}
+
+func (s *store) Append(ctx context.Context, headers ...*ExtendedHeader) error {
+	lh := len(headers)
+	if lh == 0 {
+		return nil
+	}
+
+	head, err := s.Head(ctx)
+	if err != nil {
+		return err
+	}
+
+	verified := make([]*ExtendedHeader, 0, lh)
+	for _, h := range headers {
+		err = VerifyAdjacent(head, h)
+		if err != nil {
+			log.Errorw("invalid header", "head", head.Hash(), "header", h.Hash(), "err", err)
+			break // if some headers are cryptographically valid, why not include them? Exactly, so let's include
+		}
+		verified, head = append(verified, h), h
+	}
+	if len(verified) == 0 {
+		return fmt.Errorf("header/store: no valid headers were given")
```

Comment:
Just a question because I'm not used to this pattern: did you and @renaynay agree on prefixing each error with the module?

Reply:
I don't remember us discussing this, so no.

Reply:
What would you do instead?

Reply:
I don't think the `header/store:` adds much to the error. But I see that everywhere else the errors are prefixed similarly. So it's OK to keep this consistent for now.

Reply:
I like it

Reply:
OK, at some point we should add a contribution guidline document that states that this is expected (and why).

Reply:
https://github.com/celestiaorg/celestia-node/issues/198